File: superset/utils/rison_filters.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""
Parser for Rison URL filters that converts simplified filter syntax
to Superset's adhoc_filters format.
"""

from __future__ import annotations

import logging
import re
from typing import Any, Optional, Union

import prison
from flask import has_request_context, request

logger = logging.getLogger(__name__)

# SQL identifiers permitted in URL-supplied filter columns. Restricted to
# alphanumeric/underscore (optionally schema-qualified) so OR-path SQL
# generation never interpolates an attacker-controlled identifier verbatim.
_SAFE_IDENTIFIER_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)?$")


def _safe_identifier(name: Any) -> Optional[str]:
    if isinstance(name, str) and _SAFE_IDENTIFIER_RE.match(name):
        return name
    return None


def _quote_sql_literal(value: Any) -> Optional[str]:
    """Quote a scalar value safely for SQL string interpolation.

    - Strings: single-quoted with `'` escaped as `''` (SQL standard).
    - Numeric / bool: rendered bare.
    - Anything else (None, dict, list, etc.): rejected.
    """
    if isinstance(value, bool):
        return "TRUE" if value else "FALSE"
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, str):
        return "'" + value.replace("'", "''") + "'"
    return None


class RisonFilterParser:
    """
    Parse Rison filter syntax from URL parameter 'f' and convert to adhoc_filters.

    Supports:
    - Simple equality: f=(country:USA)
    - Lists (IN): f=(country:!(USA,Canada))
    - NOT operator: f=(NOT:(country:USA))
    - OR operator: f=(OR:!(condition1,condition2))
    - Comparison operators: f=(sales:(gt:100000))
    - BETWEEN: f=(date:(between:!(2024-01-01,2024-12-31)))
    - LIKE: f=(name:(like:'%smith%'))
    """

    OPERATORS: dict[str, str] = {
        "gt": ">",
        "gte": ">=",
        "lt": "<",
        "lte": "<=",
        "between": "BETWEEN",
        "like": "LIKE",
        "ilike": "ILIKE",
        "ne": "!=",
        "eq": "==",
    }

    def parse(self, filter_string: Optional[str] = None) -> list[dict[str, Any]]:
        """
        Parse Rison filter string and convert to adhoc_filters format.

        Args:
            filter_string: Rison-encoded filter string, or None to get from request

        Returns:
            List of adhoc_filter dictionaries
        """
        if filter_string is None:
            # Callers outside a Flask request context (CLI, async tasks, tests)
            # would otherwise hit a RuntimeError on `request.args` access.
            if not has_request_context():
                return []
            filter_string = request.args.get("f")

        if not filter_string:
            return []

        try:
            filters_obj = prison.loads(filter_string)
            return self._convert_to_adhoc_filters(filters_obj)
        except Exception:
            logger.warning(
                "Failed to parse Rison filters: %s", filter_string, exc_info=True
            )
            return []

    def _convert_to_adhoc_filters(
        self, filters_obj: Union[dict[str, Any], list[Any], Any]
    ) -> list[dict[str, Any]]:
        if not isinstance(filters_obj, dict):
            return []

        adhoc_filters: list[dict[str, Any]] = []

        for key, value in filters_obj.items():
            if key == "OR":
                adhoc_filters.extend(self._handle_or_operator(value))
            elif key == "NOT":
                adhoc_filters.extend(self._handle_not_operator(value))
            else:
                filter_dict = self._create_filter(key, value)
                if filter_dict:
                    adhoc_filters.append(filter_dict)

        return adhoc_filters

    def _create_filter(
        self, column: str, value: Any, negate: bool = False
    ) -> Optional[dict[str, Any]]:
        filter_dict: dict[str, Any] = {
            "expressionType": "SIMPLE",
            "clause": "WHERE",
            "subject": column,
        }

        if isinstance(value, list):
            filter_dict["operator"] = "NOT IN" if negate else "IN"
            filter_dict["comparator"] = value
        elif isinstance(value, dict):
            operator_info = self._parse_operator_dict(value)
            if operator_info:
                operator, comparator = operator_info
                if negate and operator == "==":
                    operator = "!="
                elif negate and operator == "IN":
                    operator = "NOT IN"
                filter_dict["operator"] = operator
                filter_dict["comparator"] = comparator
            else:
                return None
        else:
            filter_dict["operator"] = "!=" if negate else "=="
            filter_dict["comparator"] = value

        return filter_dict

    def _parse_operator_dict(
        self, op_dict: dict[str, Any]
    ) -> Optional[tuple[str, Any]]:
        if not op_dict:
            return None

        for op_key, op_value in op_dict.items():
            if op_key in self.OPERATORS:
                operator = self.OPERATORS[op_key]
                if (
                    operator == "BETWEEN"
                    and isinstance(op_value, list)
                    and len(op_value) == 2
                ):
                    return operator, op_value
                return operator, op_value
            if op_key == "in":
                return "IN", op_value if isinstance(op_value, list) else [op_value]
            if op_key == "nin":
                return "NOT IN", op_value if isinstance(op_value, list) else [op_value]

        return None

    def _handle_or_operator(self, or_value: Any) -> list[dict[str, Any]]:
        if not isinstance(or_value, list):
            return []

        sql_parts: list[str] = []

        for item in or_value:
            if isinstance(item, dict):
                for col, val in item.items():
                    if col not in ("OR", "NOT"):
                        sql_part = self._build_sql_condition(col, val)
                        if sql_part:
                            sql_parts.append(sql_part)

        if sql_parts:
            return [
                {
                    "expressionType": "SQL",
                    "clause": "WHERE",
                    "sqlExpression": f"({' OR '.join(sql_parts)})",
                }
            ]

        return []

    def _build_sql_condition(self, column: str, value: Any) -> Optional[str]:
        # URL-supplied columns flow directly into a raw SQL expression in the
        # OR path, so the identifier must match a strict whitelist or we drop
        # the whole condition. Likewise, string literals get `'` escaped to
        # `''` and unrecognised value types are rejected outright.
        safe_column = _safe_identifier(column)
        if safe_column is None:
            logger.warning("Rejecting URL filter with unsafe column: %r", column)
            return None

        if isinstance(value, list):
            quoted = [_quote_sql_literal(v) for v in value]
            if any(q is None for q in quoted):
                return None
            return f"{safe_column} IN ({', '.join(quoted)})"  # type: ignore[arg-type]

        if isinstance(value, dict):
            operator_info = self._parse_operator_dict(value)
            if operator_info:
                op, comp = operator_info
                if op == "BETWEEN" and isinstance(comp, list) and len(comp) == 2:
                    lo = _quote_sql_literal(comp[0])
                    hi = _quote_sql_literal(comp[1])
                    if lo is None or hi is None:
                        return None
                    return f"{safe_column} BETWEEN {lo} AND {hi}"
                comp_str = _quote_sql_literal(comp)
                if comp_str is None:
                    return None
                return f"{safe_column} {op} {comp_str}"
            return None

        val_str = _quote_sql_literal(value)
        if val_str is None:
            return None
        return f"{safe_column} = {val_str}"

    def _handle_not_operator(self, not_value: Any) -> list[dict[str, Any]]:
        if isinstance(not_value, dict):
            filters: list[dict[str, Any]] = []
            for col, val in not_value.items():
                if col not in ("OR", "NOT"):
                    filter_dict = self._create_filter(col, val, negate=True)
                    if filter_dict:
                        filters.append(filter_dict)
            return filters

        return []


def merge_rison_filters(form_data: dict[str, Any]) -> None:
    """
    Merge Rison filters from 'f' parameter into form_data.
    Modifies form_data in place.
    """
    parser = RisonFilterParser()

    if rison_filters := parser.parse():
        existing_filters = form_data.get("adhoc_filters", [])
        form_data["adhoc_filters"] = existing_filters + rison_filters
        logger.info("Added %d filters from Rison parameter", len(rison_filters))

File: superset-frontend/src/dashboard/util/risonFilters.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  QueryObjectFilterClause,
  PartialFilters,
  DataMaskStateWithId,
} from '@superset-ui/core';
import rison from 'rison';
import { navigateWithState } from 'src/utils/navigationUtils';

/**
 * Synthetic dataMask key for URL Rison filters that don't match any native
 * filter on the dashboard. Because no entry in `nativeFilters` claims this id,
 * `getAllActiveFilters` falls through to `allSliceIds`, so the attached
 * `extraFormData.filters` apply to every chart on the dashboard.
 */
export const RISON_UNMATCHED_DATAMASK_ID = '__rison_unmatched__';

export interface RisonFilter {
  subject: string;
  operator: string;
  comparator: string | number | boolean | (string | number)[];
}

export interface IntelligentRisonInjectionResult {
  updatedDataMask: DataMaskStateWithId;
  unmatchedFilters: RisonFilter[];
}

/**
 * Parse individual filter condition
 */
function parseFilterCondition(key: string, value: unknown): RisonFilter {
  // Handle comparison operators: (gt:100), (between:!(1,10))
  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
    const [operator, operatorValue] = Object.entries(
      value as Record<string, unknown>,
    )[0];

    switch (operator) {
      case 'gt':
        return {
          subject: key,
          operator: '>',
          comparator: operatorValue as string | number,
        };
      case 'gte':
        return {
          subject: key,
          operator: '>=',
          comparator: operatorValue as string | number,
        };
      case 'lt':
        return {
          subject: key,
          operator: '<',
          comparator: operatorValue as string | number,
        };
      case 'lte':
        return {
          subject: key,
          operator: '<=',
          comparator: operatorValue as string | number,
        };
      case 'between':
        return {
          subject: key,
          operator: 'BETWEEN',
          comparator: operatorValue as (string | number)[],
        };
      case 'like':
        return {
          subject: key,
          operator: 'LIKE',
          comparator: operatorValue as string,
        };
      default:
        return {
          subject: key,
          operator: '==',
          comparator: value as unknown as string | number,
        };
    }
  }

  // Handle IN operator: !(value1,value2)
  if (Array.isArray(value)) {
    return {
      subject: key,
      operator: 'IN',
      comparator: value as (string | number)[],
    };
  }

  // Handle simple equality
  return {
    subject: key,
    operator: '==',
    comparator: value as string | number | boolean,
  };
}

/**
 * Parse Rison filter syntax from URL parameter.
 * Supports formats like: (country:USA,year:2024)
 */
export function parseRisonFilters(risonString: string): RisonFilter[] {
  try {
    const parsed = rison.decode(risonString);
    const filters: RisonFilter[] = [];

    if (!parsed || typeof parsed !== 'object') {
      return filters;
    }

    const parsedObj = parsed as Record<string, unknown>;

    // Handle OR operator: OR:!(condition1,condition2)
    if (parsedObj.OR && Array.isArray(parsedObj.OR)) {
      (parsedObj.OR as Record<string, unknown>[]).forEach(condition => {
        if (typeof condition === 'object') {
          Object.entries(condition).forEach(([key, value]) => {
            filters.push(parseFilterCondition(key, value));
          });
        }
      });
      return filters;
    }

    // Handle NOT operator: NOT:(condition). Falls through so regular keys at the
    // same level are still picked up below (supports mixed payloads like
    // `(country:USA,NOT:(status:archived))`).
    if (parsedObj.NOT && typeof parsedObj.NOT === 'object') {
      Object.entries(parsedObj.NOT as Record<string, unknown>).forEach(
        ([key, value]) => {
          const filter = parseFilterCondition(key, value);
          if (filter.operator === '==') {
            filter.operator = '!=';
          } else if (filter.operator === 'IN') {
            filter.operator = 'NOT IN';
          }
          filters.push(filter);
        },
      );
    }

    // Handle regular filters
    Object.entries(parsedObj).forEach(([key, value]) => {
      if (key !== 'OR' && key !== 'NOT') {
        filters.push(parseFilterCondition(key, value));
      }
    });

    return filters;
  } catch (error) {
    console.warn('Failed to parse Rison filters:', error);
    return [];
  }
}

/**
 * Convert Rison filters to Superset adhoc filter format
 */
export function risonToAdhocFilters(
  risonFilters: RisonFilter[],
): QueryObjectFilterClause[] {
  return risonFilters.map(
    filter =>
      ({
        expressionType: 'SIMPLE' as const,
        clause: 'WHERE' as const,
        subject: filter.subject,
        operator: filter.operator,
        comparator: filter.comparator,
      }) as unknown as QueryObjectFilterClause,
  );
}

/**
 * Prettify Rison filter URL by replacing encoded characters.
 * Uses browser history API to update URL without page reload.
 */
export function prettifyRisonFilterUrl(): void {
  try {
    const currentUrl = window.location.href;

    if (!currentUrl.includes('&f=') && !currentUrl.includes('?f=')) {
      return;
    }

    const urlMatch = currentUrl.match(/([?&])f=([^&]*)/);
    if (!urlMatch) {
      return;
    }

    const separator = urlMatch[1];
    let risonValue = urlMatch[2];

    if (!risonValue.includes('%') && !risonValue.includes('+')) {
      return;
    }

    let previousValue = '';
    let decodeAttempts = 0;
    while (risonValue !== previousValue && decodeAttempts < 5) {
      previousValue = risonValue;
      try {
        if (risonValue.includes('%')) {
          risonValue = decodeURIComponent(risonValue);
        }
      } catch {
        break;
      }
      decodeAttempts += 1;
    }

    risonValue = risonValue.replace(/\+/g, ' ');

    const matchIndex = urlMatch.index ?? 0;
    const beforeRison = currentUrl.substring(0, matchIndex);
    const afterRison = currentUrl.substring(matchIndex + urlMatch[0].length);
    const prettifiedUrl = `${beforeRison}${separator}f=${risonValue}${afterRison}`;

    if (prettifiedUrl !== currentUrl) {
      // Route through navigateWithState so the navigationUtils guards
      // (`assertSafeNavigationUrl` scheme/userinfo barriers + the
      // CodeQL-recognised inline sanitisers) apply at the
      // `window.history.replaceState` sink. The URL constructor inside
      // `navigateWithState` is conservative about re-encoding: sub-delims
      // like `(`, `)`, `:`, `!` (the meaningful Rison glyphs) survive,
      // so the prettification's visual win is preserved for every
      // character the prettifier actually targets.
      navigateWithState(prettifiedUrl, window.history.state ?? {}, {
        replace: true,
      });
    }
  } catch (error) {
    console.warn('Failed to prettify Rison URL:', error);
  }
}

/**
 * Get Rison filter parameter from URL
 */
export function getRisonFilterParam(): string | null {
  const params = new URLSearchParams(window.location.search);
  return params.get('f');
}

/**
 * Convert an array of RisonFilter back to Rison string format
 */
export function risonFiltersToString(filters: RisonFilter[]): string {
  if (filters.length === 0) {
    return '';
  }

  type RisonValue =
    string | number | boolean | (string | number)[] | Record<string, unknown>;
  const risonObject: Record<string, RisonValue> = {};
  const notObject: Record<string, RisonValue> = {};

  const encodePositive = (
    target: Record<string, RisonValue>,
    filter: RisonFilter,
    op: string,
  ) => {
    if (op === 'IN' && Array.isArray(filter.comparator)) {
      target[filter.subject] = filter.comparator;
    } else if (op === '==') {
      target[filter.subject] = filter.comparator;
    } else {
      const operatorMap: Record<string, string> = {
        '>': 'gt',
        '>=': 'gte',
        '<': 'lt',
        '<=': 'lte',
        BETWEEN: 'between',
        LIKE: 'like',
        ILIKE: 'ilike',
      };
      const risonOp = operatorMap[op] || op;
      target[filter.subject] = { [risonOp]: filter.comparator };
    }
  };

  filters.forEach(filter => {
    if (filter.operator === '!=') {
      // Re-emit as NOT:(col:value) so parseRisonFilters can read it back.
      encodePositive(notObject, filter, '==');
    } else if (filter.operator === 'NOT IN') {
      encodePositive(notObject, filter, 'IN');
    } else {
      encodePositive(risonObject, filter, filter.operator);
    }
  });

  if (Object.keys(notObject).length > 0) {
    risonObject.NOT = notObject;
  }

  try {
    return rison.encode(risonObject);
  } catch (error) {
    console.warn('Failed to encode Rison filters:', error);
    return '';
  }
}

interface ReplaceHistory {
  replace(location: { pathname: string; search: string }): void;
}

/**
 * Update the URL to remove successfully matched filters, keeping only unmatched ones.
 * When a react-router history is supplied, the update goes through it so that
 * components reading from `history.location` (e.g. `publishDataMask` in the
 * filter bar) see the new search string. Otherwise falls back to a raw
 * `window.history.replaceState`.
 */
export function updateUrlWithUnmatchedFilters(
  unmatchedFilters: RisonFilter[],
  history?: ReplaceHistory,
): void {
  try {
    const currentUrl = new URL(window.location.href);

    if (unmatchedFilters.length === 0) {
      currentUrl.searchParams.delete('f');
    } else {
      const newRisonString = risonFiltersToString(unmatchedFilters);
      if (newRisonString) {
        currentUrl.searchParams.set('f', newRisonString);
      } else {
        currentUrl.searchParams.delete('f');
      }
    }

    // Always keep window.history in sync so callers that read
    // `window.location.search` (e.g. `getRisonFilterParam`) see the update.
    // With a real `BrowserRouter`, `history.replace` would do this too — but
    // under a `createMemoryHistory` (used in tests, or in some embedded
    // contexts) it does not, and we'd leak the stale URL into the next
    // `getRisonFilterParam()` call. Routed through navigateWithState so the
    // navigationUtils scheme/userinfo barriers gate the sink.
    navigateWithState(currentUrl.toString(), window.history.state ?? {}, {
      replace: true,
    });
    if (history) {
      history.replace({
        pathname: currentUrl.pathname,
        search: currentUrl.search,
      });
    }
  } catch (error) {
    console.warn('Failed to update URL with unmatched filters:', error);
  }
}

/**
 * Find a native filter that matches a Rison filter by column name.
 * Uses case-insensitive, trimmed comparison to handle column names with spaces.
 */
function findMatchingNativeFilter(
  risonFilter: RisonFilter,
  nativeFilters: PartialFilters,
): string | null {
  const normalizedSubject = risonFilter.subject.trim().toLowerCase();

  for (const [filterId, nativeFilter] of Object.entries(nativeFilters)) {
    if (!nativeFilter?.targets) continue;

    const hasMatchingTarget = nativeFilter.targets.some(target => {
      if (typeof target === 'object' && target && 'column' in target) {
        return target.column?.name?.trim().toLowerCase() === normalizedSubject;
      }
      return false;
    });

    if (hasMatchingTarget) {
      return filterId;
    }
  }

  return null;
}

/**
 * Build extraFormData filters for a given rison filter and column name
 */
function buildExtraFormDataFilters(
  risonFilter: RisonFilter,
  columnName: string,
): QueryObjectFilterClause[] {
  const { operator, comparator } = risonFilter;

  if (operator === 'IN' || (operator === '==' && Array.isArray(comparator))) {
    return [
      {
        col: columnName,
        op: 'IN',
        val: Array.isArray(comparator) ? comparator : [comparator],
      },
    ];
  }

  if (operator === '==' && !Array.isArray(comparator)) {
    return [{ col: columnName, op: 'IN', val: [comparator] }];
  }

  if (
    operator === 'BETWEEN' &&
    Array.isArray(comparator) &&
    comparator.length === 2
  ) {
    return [
      { col: columnName, op: '>=', val: comparator[0] },
      { col: columnName, op: '<=', val: comparator[1] },
    ];
  }

  return [
    {
      col: columnName,
      op: operator,
      val: comparator,
    } as QueryObjectFilterClause,
  ];
}

/**
 * Convert a list of Rison filters into the `{col, op, val}` clauses expected
 * by `dataMask[id].extraFormData.filters`. Each filter uses its own subject
 * as the column name.
 */
export function risonFiltersToExtraFormDataFilters(
  filters: RisonFilter[],
): QueryObjectFilterClause[] {
  return filters.flatMap(filter =>
    buildExtraFormDataFilters(filter, filter.subject),
  );
}

/**
 * Convert a Rison filter value to the format expected by a native filter.
 * Also returns extraFormData for auto-application.
 */
function convertRisonToNativeValue(
  risonFilter: RisonFilter,
  nativeFilter: { filterType?: string },
): unknown {
  const { comparator, operator } = risonFilter;
  const filterType = nativeFilter?.filterType;

  switch (filterType) {
    case 'filter_select':
      if (operator === 'IN' || Array.isArray(comparator)) {
        return Array.isArray(comparator) ? comparator : [comparator];
      }
      return [comparator];

    case 'filter_range':
      if (
        operator === 'BETWEEN' &&
        Array.isArray(comparator) &&
        comparator.length === 2
      ) {
        return { min: comparator[0], max: comparator[1] };
      }
      return comparator;

    case 'filter_time_range':
    case 'filter_timecolumn':
      return comparator;

    default:
      return Array.isArray(comparator) ? comparator : [comparator];
  }
}

/**
 * Build a complete DataMask entry for a rison filter matched to a native filter.
 * Sets both filterState.value AND extraFormData so the filter auto-applies.
 */
function buildDataMaskForFilter(
  risonFilter: RisonFilter,
  nativeFilter: {
    id: string;
    filterType?: string;
    targets?: { column?: { name?: string } }[];
  },
  columnName: string,
) {
  const convertedValue = convertRisonToNativeValue(risonFilter, nativeFilter);

  return {
    id: nativeFilter.id,
    filterState: {
      value: convertedValue,
    },
    extraFormData: {
      filters: buildExtraFormDataFilters(risonFilter, columnName),
    },
    ownState: {},
  };
}

/**
 * Intelligently inject Rison filters into native filters where possible,
 * falling back to brute-force injection for unmatched filters
 */
export function injectRisonFiltersIntelligently(
  risonFilters: RisonFilter[],
  nativeFilters: PartialFilters,
  currentDataMask: DataMaskStateWithId,
): IntelligentRisonInjectionResult {
  const updatedDataMask = { ...currentDataMask };
  const unmatchedFilters: RisonFilter[] = [];

  risonFilters.forEach(risonFilter => {
    const matchingFilterId = findMatchingNativeFilter(
      risonFilter,
      nativeFilters,
    );

    if (matchingFilterId) {
      const matchedFilter = nativeFilters[matchingFilterId];
      const filterId = matchedFilter?.id ?? matchingFilterId;
      if (matchedFilter && filterId) {
        const columnName =
          matchedFilter.targets?.[0]?.column?.name ?? risonFilter.subject;

        const dataMaskEntry = buildDataMaskForFilter(
          risonFilter,
          { ...matchedFilter, id: filterId } as {
            id: string;
            filterType?: string;
            targets?: { column?: { name?: string } }[];
          },
          columnName,
        );

        updatedDataMask[filterId] = {
          ...updatedDataMask[filterId],
          ...dataMaskEntry,
        };
        return;
      }
    }

    unmatchedFilters.push(risonFilter);
  });

  return {
    updatedDataMask,
    unmatchedFilters,
  };
}
